ledger/eval.go

```diff
@@ -479,10 +535,15 @@ func (eval *BlockEvaluator) transactionGroup(txgroup []transactions.SignedTxnWit
 
 	cow := eval.state.child()
 
+	// Prepare TEAL contexts for any ApplicationCall transactions in the group
+	appEvaluators := eval.prepareAppEvaluators(txgroup)
```

Comment:
Here's my understanding of the current mechanisms to expose ledger state to `logic.Eval`:

1. From `cow.TransactionGroup`, prepare a `appTealEvaluator` for each transaction in the group. Pass this to `Transaction.Apply` via the `transactions.StateEvaluator` interface. `Transaction.Apply` calls `InitLedger`, which initializes the `StateEvaluator` so that it correctly implements `ledger.LedgerForLogic` (specifically by creating an `appLedger`), and then calls `StateEvaluator.Check` and `StateEvaluator.Apply`, which internally pass the `appLedger`, which implements `logic.LedgerForLogic` to `logic.Eval`.

2. From the TEAL debugger, call `MakeDebugAppLedger`, which returns an `appLedger`, and pass this directly to `logic.Eval`.

Is the above correct?

Reply:
If so, have you considered instead preparing an `appLedger` instead of an `appTealEvaluator` for each transaction in the group, dropping `InitLedger` (so others don't need to remember to call it), and having `appLedger` implement both `StateEvaluator` and `LedgerForLogic`?

Reply:
(Here's a mediocre diagram showing the proposed changes, which are illustrated with blue pen:)

[https://drive.google.com/file/d/1z11-UdQUXsns-f1gOuXs-nBY1qJYdlfj/view?usp=sharing](url)

Reply:
I think your understanding of the code is correct.

The reason `InitLedger` exists is that if it didn't, the ledger would need to know more about application internals. In particular, `logic.Eval` needs to know about the application ID even in the event that the application is being created by this transaction (for use in the access whitelist), and it needs to know about the application's state schemas even in the event that the application is _not_ being created (which requires looking up the creator's balance record).

In practice, I think this would mean putting logic in the ledger that 1. allocates application IDs using the transaction counter and 2. looks up existing application parameters before calling `apply` OR uses the schemas mentioned in the transaction if this is a creation txn. If we're OK doing that, then we can make that change. I think it would make the interface cleaner in exchange for moving application specific details into the ledger. What do you think?

Reply:
Ah, okay, makes sense. Yeah, it might be better to keep the current design, in that case.

Reply:
I have simplified `InitLedger` substantially by removing unnecessary whitelisting here: https://github.com/justicz/go-algorand/pull/153 . This makes `ledger/application.go` a significantly thinner wrapper around balances.